File: copilot-bot/splitsend/splitsend.go
package splitsend

import (
	"regexp"
	"strings"
)

const maxLength = 2000

// markdownPair represents a markdown delimiter pattern that comes in pairs.
type markdownPair struct {
	pattern *regexp.Regexp
	name    string
}

var markdownPairs = []markdownPair{
	{regexp.MustCompile("```"), "code block"},
	{regexp.MustCompile("``"), "inline code"},
	{regexp.MustCompile("`"), "inline code"},
	{regexp.MustCompile(`\*\*`), "bold"},
	{regexp.MustCompile(`__`), "underline bold"},
	{regexp.MustCompile(`\|\|`), "spoiler"},
	{regexp.MustCompile(`~~`), "strikethrough"},
	{regexp.MustCompile(`(?:^|[^*])\*(?:[^*]|$)`), "italic"},
	{regexp.MustCompile(`(?:^|[^_])_(?:[^_]|$)`), "italic"},
}

// countMatches counts occurrences of a pattern in a string.
func countMatches(text string, pattern *regexp.Regexp) int {
	return len(pattern.FindAllStringIndex(text, -1))
}

// hasUnclosedMarkdown checks if a chunk has unclosed markdown formatting.
func hasUnclosedMarkdown(chunk string) bool {
	for _, pair := range markdownPairs {
		count := countMatches(chunk, pair.pattern)
		if count%2 != 0 {
			return true
		}
	}
	return false
}

// findSafeSplitPoint finds the last safe split point that doesn't break markdown formatting.
func findSafeSplitPoint(text string, maxLen int) int {
	if maxLen > len(text) {
		maxLen = len(text)
	}

	// Try to split at a newline for cleaner breaks
	lastNewline := strings.LastIndex(text[:maxLen], "\n")
	if lastNewline > maxLen/2 {
		chunk := text[:lastNewline]
		if !hasUnclosedMarkdown(chunk) {
			return lastNewline + 1 // Include the newline in the first chunk
		}
	}

	// Try splitting at paragraph breaks (double newlines)
	lastParagraph := strings.LastIndex(text[:maxLen], "\n\n")
	if lastParagraph > maxLen*3/10 {
		chunk := text[:lastParagraph]
		if !hasUnclosedMarkdown(chunk) {
			return lastParagraph + 2
		}
	}

	// Try splitting at code block boundaries
	codeBlockRegex := regexp.MustCompile("(?s)```.*?```")
	textToSearch := text[:maxLen]
	matches := codeBlockRegex.FindAllStringIndex(textToSearch, -1)
	lastCodeBlockEnd := 0
	for _, match := range matches {
		endPos := match[1]
		if endPos <= maxLen {
			lastCodeBlockEnd = endPos
		}
	}
	if lastCodeBlockEnd > maxLen*3/10 {
		return lastCodeBlockEnd
	}

	// Work backwards from maxLen, preferring spaces/newlines with valid markdown closure
	for i := maxLen; i > maxLen*3/10; i-- {
		if i >= len(text) {
			continue
		}
		ch := text[i]
		if ch == '\n' || ch == ' ' {
			chunk := text[:i]
			if !hasUnclosedMarkdown(chunk) {
				if ch == '\n' {
					return i + 1
				}
				return i
			}
		}
	}

	// Fall back to splitting at a space
	lastSpace := strings.LastIndex(text[:maxLen], " ")
	if lastSpace > maxLen*3/10 {
		return lastSpace
	}

	// Last resort: hard split at maxLen
	return maxLen
}

// SplitMessage splits a message into chunks respecting markdown formatting boundaries
// and Discord's 2000-character limit.
func SplitMessage(message string) []string {
	var chunks []string
	remaining := message

	for len(remaining) > 0 {
		if len(remaining) <= maxLength {
			chunks = append(chunks, remaining)
			break
		}

		splitPoint := findSafeSplitPoint(remaining, maxLength)
		chunk := strings.TrimSpace(remaining[:splitPoint])

		if len(chunk) > 0 {
			chunks = append(chunks, chunk)
		}

		remaining = strings.TrimSpace(remaining[splitPoint:])
	}

	return chunks
}

File: copilot-bot-swift/Sources/CopilotBot/SplitSend.swift
import Foundation

private let maxLength = 2000

private struct MarkdownPair {
    let pattern: NSRegularExpression
    let name: String
}

private let markdownPairs: [MarkdownPair] = [
    makePattern("```", name: "code block"),
    makePattern("``", name: "inline code 2"),
    makePattern("`", name: "inline code"),
    makePattern(#"\*\*"#, name: "bold"),
    makePattern("__", name: "underline bold"),
    makePattern(#"\|\|"#, name: "spoiler"),
    makePattern("~~", name: "strikethrough"),
    makePattern(#"(?:^|[^*])\*(?:[^*]|$)"#, name: "italic *"),
    makePattern(#"(?:^|[^_])_(?:[^_]|$)"#, name: "italic _"),
]

private func makePattern(_ pattern: String, name: String) -> MarkdownPair {
    let regex = try! NSRegularExpression(pattern: pattern, options: [.anchorsMatchLines])
    return MarkdownPair(pattern: regex, name: name)
}

private func countMatches(in text: String, pattern: NSRegularExpression) -> Int {
    pattern.numberOfMatches(in: text, range: NSRange(text.startIndex..., in: text))
}

private func hasUnclosedMarkdown(_ chunk: String) -> Bool {
    for pair in markdownPairs {
        if countMatches(in: chunk, pattern: pair.pattern) % 2 != 0 {
            return true
        }
    }
    return false
}

private func findSafeSplitPoint(in text: String, maxLen: Int) -> String.Index {
    let clampedEnd = text.index(text.startIndex, offsetBy: min(maxLen, text.count))
    let prefix = String(text[..<clampedEnd])

    // Try newline
    if let idx = prefix.range(of: "\n", options: .backwards)?.lowerBound,
       text.distance(from: text.startIndex, to: idx) > maxLen / 2 {
        let chunk = String(text[..<idx])
        if !hasUnclosedMarkdown(chunk) {
            return text.index(after: idx) // skip the newline itself
        }
    }

    // Try paragraph break
    if let idx = prefix.range(of: "\n\n", options: .backwards)?.lowerBound,
       text.distance(from: text.startIndex, to: idx) > maxLen * 3 / 10 {
        let chunk = String(text[..<idx])
        if !hasUnclosedMarkdown(chunk) {
            return text.index(idx, offsetBy: 2)
        }
    }

    // Scan backwards for space/newline
    var i = clampedEnd
    let lowerBound = text.index(text.startIndex, offsetBy: maxLen * 3 / 10)
    while i > lowerBound {
        text.formIndex(before: &i)
        let ch = text[i]
        if ch == "\n" || ch == " " {
            let chunk = String(text[..<i])
            if !hasUnclosedMarkdown(chunk) {
                return ch == "\n" ? text.index(after: i) : i
            }
        }
    }

    // Last space
    if let idx = prefix.range(of: " ", options: .backwards)?.lowerBound,
       text.distance(from: text.startIndex, to: idx) > maxLen * 3 / 10 {
        return idx
    }

    // Hard cut
    return clampedEnd
}

/// Splits a message into chunks of at most 2000 characters, respecting markdown boundaries.
func splitMessage(_ message: String) -> [String] {
    var chunks: [String] = []
    var remaining = message

    while !remaining.isEmpty {
        if remaining.count <= maxLength {
            chunks.append(remaining)
            break
        }
        let splitIndex = findSafeSplitPoint(in: remaining, maxLen: maxLength)
        let chunk = String(remaining[..<splitIndex]).trimmingCharacters(in: .whitespaces)
        if !chunk.isEmpty {
            chunks.append(chunk)
        }
        remaining = String(remaining[splitIndex...]).trimmingCharacters(in: .whitespaces)
    }

    return chunks
}
